File: android/app/src/main/kotlin/com/endurain/endurain/activity/ActivityRecorderCoordinator.kt
package com.endurain.endurain.activity

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * In-process bridge between the foreground service and the Flutter event sink.
 *
 * The foreground service and the [ActivityRecorderChannel] run in the same
 * process, so a volatile sink reference is sufficient. When Flutter is detached
 * (app killed/backgrounded with no engine) events are dropped silently; points
 * are still persisted durably and replayed on the next `drain`.
 *
 * Payloads carry only sanitized lifecycle data and track points; no file paths,
 * tokens, or free-form diagnostics are emitted here.
 */
object ActivityRecorderCoordinator {
    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    private val mainHandler = Handler(Looper.getMainLooper())

    fun attach(sink: EventChannel.EventSink) {
        eventSink = sink
    }

    fun detach() {
        eventSink = null
    }

    fun emitSession(type: String, session: ActiveActivitySessionData) {
        emit(
            mapOf(
                KEY_TYPE to type,
                KEY_SESSION to session.toMap(),
            ),
        )
    }

    fun emitPointBatch(points: List<RecordedActivityPointData>) {
        if (points.isEmpty()) {
            return
        }
        emit(
            mapOf(
                KEY_TYPE to TYPE_POINT_BATCH_AVAILABLE,
                KEY_POINTS to points.map { it.toMap() },
            ),
        )
    }

    fun emitRecoverableStateChanged(session: ActiveActivitySessionData?) {
        val payload = LinkedHashMap<String, Any?>()
        payload[KEY_TYPE] = TYPE_RECOVERABLE_STATE_CHANGED
        if (session != null) {
            payload[KEY_SESSION] = session.toMap()
        }
        emit(payload)
    }

    fun emitFailed(reason: String) {
        emit(
            mapOf(
                KEY_TYPE to TYPE_FAILED,
                KEY_REASON to reason,
            ),
        )
    }

    private fun emit(payload: Map<String, Any?>) {
        val sink = eventSink ?: return
        mainHandler.post {
            // The sink may have detached between the null check and delivery.
            if (eventSink === sink) {
                sink.success(payload)
            }
        }
    }

    // Event payload keys (mirror NativeActivityRecorderChannelContract).
    const val KEY_TYPE = "type"
    const val KEY_SESSION = "session"
    const val KEY_POINTS = "points"
    const val KEY_REASON = "reason"

    // Event type values.
    const val TYPE_STARTED = "started"
    const val TYPE_POINT_BATCH_AVAILABLE = "pointBatchAvailable"
    const val TYPE_PAUSED = "paused"
    const val TYPE_RESUMED = "resumed"
    const val TYPE_STOPPED = "stopped"
    const val TYPE_FAILED = "failed"
    const val TYPE_RECOVERABLE_STATE_CHANGED = "recoverableStateChanged"

    // Failure reasons (mirror Dart ActivityRecorderFailureReason names).
    const val REASON_LOCATION_STREAM_FAILED = "locationStreamFailed"
    const val REASON_LOCATION_UNAVAILABLE = "locationUnavailable"
    const val REASON_PERMISSION_LOST = "permissionLost"
    const val REASON_PERSISTENCE_FAILED = "persistenceFailed"
    const val REASON_UNSUPPORTED_PLATFORM = "unsupportedPlatform"
}

File: ios/Runner/Activity/ActivityRecorderCoordinator.swift
import Foundation
import Flutter

/// Bridges native recorder events to the Flutter `EventChannel` sink.
///
/// Mirrors the Android `ActivityRecorderCoordinator`. Event payloads are
/// versioned and use the same `type`/`session`/`points`/`reason` keys the Dart
/// channel parser expects. All emissions are marshalled to the main thread,
/// which is where the platform channel expects to be driven. Never emits raw
/// coordinates in diagnostics; only structured, channel-bound payloads.
final class ActivityRecorderCoordinator {
    static let shared = ActivityRecorderCoordinator()

    private let payloadVersion = 1

    private var eventSink: FlutterEventSink?
    private var eventSinkGeneration = 0

    private init() {}

    func attach(_ sink: @escaping FlutterEventSink) {
        runOnMain { [weak self] in
            self?.eventSinkGeneration += 1
            self?.eventSink = sink
        }
    }

    func detach() {
        runOnMain { [weak self] in
            self?.eventSinkGeneration += 1
            self?.eventSink = nil
        }
    }

    func emitSession(type: String, session: ActiveActivitySessionData) {
        emit([
            "version": payloadVersion,
            "type": type,
            "session": session.toMap(),
        ])
    }

    func emitPointBatch(_ points: [RecordedActivityPointData]) {
        if points.isEmpty {
            return
        }
        emit([
            "version": payloadVersion,
            "type": ActivityRecorderCoordinator.eventPointBatchAvailable,
            "points": points.map { $0.toMap() },
        ])
    }

    func emitRecoverableStateChanged(_ session: ActiveActivitySessionData?) {
        var payload: [String: Any] = [
            "version": payloadVersion,
            "type": ActivityRecorderCoordinator.eventRecoverableStateChanged,
        ]
        if let session = session {
            payload["session"] = session.toMap()
        }
        emit(payload)
    }

    func emitFailed(_ reason: String) {
        emit([
            "version": payloadVersion,
            "type": ActivityRecorderCoordinator.eventFailed,
            "reason": reason,
        ])
    }

    private func emit(_ payload: [String: Any]) {
        if Thread.isMainThread {
            eventSink?(payload)
        } else {
            var capturedGeneration = 0
            var hasSink = false
            DispatchQueue.main.sync { [weak self] in
                capturedGeneration = self?.eventSinkGeneration ?? 0
                hasSink = self?.eventSink != nil
            }
            guard hasSink else {
                return
            }
            DispatchQueue.main.async { [weak self] in
                guard
                    let self = self,
                    self.eventSinkGeneration == capturedGeneration
                else {
                    return
                }
                self.eventSink?(payload)
            }
        }
    }

    private func runOnMain(_ work: @escaping () -> Void) {
        if Thread.isMainThread {
            work()
        } else {
            DispatchQueue.main.async(execute: work)
        }
    }

    static let eventStarted = "started"
    static let eventPointBatchAvailable = "pointBatchAvailable"
    static let eventPaused = "paused"
    static let eventResumed = "resumed"
    static let eventAutoPaused = "autoPaused"
    static let eventAutoResumed = "autoResumed"
    static let eventStopped = "stopped"
    static let eventFailed = "failed"
    static let eventRecoverableStateChanged = "recoverableStateChanged"

    static let reasonLocationStreamFailed = "locationStreamFailed"
    static let reasonLocationUnavailable = "locationUnavailable"
    static let reasonPermissionLost = "permissionLost"
    static let reasonPersistenceFailed = "persistenceFailed"
    static let reasonUnsupportedPlatform = "unsupportedPlatform"
}
